helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -1134,7 +1163,17 @@ public void handleNewSession(final String sessionId) throws Exception {
     // Wait until we get a non-zero session id. Otherwise, getSessionId() might be null.
     waitUntilConnected();
 
-    // TODO: filter out stale sessions here.
+    /*
+     * Filter out stale sessions. If a session id is not null and not the same as current session
+     * id, this session is expired. With this filtering, expired sessions are NOT handled,
+     * so performance is expected to improve.
+     */
+    if (sessionId != null && !getSessionId().equals(sessionId)) {
```

Comment:
what happens if sessionId is null, will all following calls  function correctly if it is null? say can createLiveInstance() handle it properly if sessionId passed to it is null?

Reply:
Like what I mention in the comment `TODO: after removing I0ItecIZkStateListenerHelixImpl, null session should be checked and discarded.`, null is acceptable for backward compatibility. If null is passed in, we will keep the behavior as before, which may cause race condition.